`numpy` is one of the most important Python packages in science. It implements a new data type, called `ndarray` (*n-dimensional array*), which is optimized for numerical operations. Here are some things you need to know about arrays:

1. In arrays all elements must have the same type and hence occupy the same memory space. You cannot define an array in which one element is an `int` and another is a `float` or `bool`.
2. Arrays occupy continous segment of memory, as opposed to lists, which are just pointers to different objects in various part of memory
3. Arrays have *constant access time*. If you have a very large `list`, getting elements from it will be progressively more slow. This is not the case with arrays: getting elements is always fast. This is a consequence of the first 2 properties of arrays.
4. However, insertion of elements in the middle of an array or appending an array is inefficient. Hence, you should always *preallocate* an array, if you can. For example, if you want to make a loop and add a value to an array on each iteration, it is much more efficient to first define an empty array which has the length of the final result, and then change elements on each iteration of the loop. We will see an example in practice later.

First let's import `numpy`. It is customary to import it as `np`.

In [1]:
import numpy as np

Simplest thing you can do is to make an array from a list, like so:

In [2]:
x = np.array([1,2,3])
print(x)

[1 2 3]


However, there are many other ways of creating arrays, `numpy` package includes a lot of functions which create arrays based of various rules, and we will use some of them throughout this notebook. One of them is `np.zeros(n)` which creates an array of length `n` filled with `0`:

In [3]:
x = np.zeros(4)
print(x)

[0. 0. 0. 0.]


When you create an array, it has some very helpful attributes, such as 
- `ndim` (number of dimensions)
- `shape` (length of each dimension)
- `size` (total number of elements in the array)
- `dtype` (element type)

In [4]:
x = np.zeros(10)
print(x)
print("ndim: ", x.ndim)
print("shape:", x.shape)
print("size: ", x.size)
print("dtype:", x.dtype)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
ndim:  1
shape: (10,)
size:  10
dtype: float64


Let's see the same for 2-dimensional array (we will take a look at how to create and work with them in more detail later):

In [5]:
x = np.zeros((5,5))
print(x)
print("ndim: ", x.ndim)
print("shape:", x.shape)
print("size: ", x.size)
print("dtype:", x.dtype)

[[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]
ndim:  2
shape: (5, 5)
size:  25
dtype: float64


Another useful function to create arrays is `np.arange(n)` which will create an array of integers from `0` to `n` (not including `n`):

In [6]:
x = np.arange(20)
x

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19])

You can also specify beginning and end `np.arange(n,m)` to created an array from `n` to `m` (excluding `m`):

In [7]:
x = np.arange(10,30)
x

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29])

Finally, you can specify a step:

In [8]:
x = np.arange(10,20,0.5)
x

array([10. , 10.5, 11. , 11.5, 12. , 12.5, 13. , 13.5, 14. , 14.5, 15. ,
       15.5, 16. , 16.5, 17. , 17.5, 18. , 18.5, 19. , 19.5])

# Array indexing

Array indexing works exactly the same way as `list` indexing. As always in Python, we start counting from zero. `x[0]` is the first element of the array, `x[3]` gives us *4-th* element of the array, etc:

In [9]:
x[0]

10.0

In [10]:
x[3]

11.5

In [11]:
x[5]

12.5

In [12]:
# indexing from the end: -1 is the last element
x[-1]

19.5

In [13]:
x[-5]

17.5

Let's create a two-dimensional array using another function from `numpy` -- `np.random.randint`, which will give as an array of random integers up to a certain number. We can also specify the size of the array with *keyword argument* `size`:

In [14]:
# create two-dimensional array of size (3,5) filled with random integers from 10 to 50
x2 = np.random.randint(10, 50, size=(3,5))
x2

array([[42, 32, 48, 37, 47],
       [28, 36, 43, 41, 39],
       [21, 31, 29, 13, 37]])

When indexing multidimensional array, specify indexes in each dimension, like so:

In [15]:
# get element which is in position 0 in first dimension, and position 4 in second dimension
x2[0,4]

47

In [16]:
# get the same element using indexing from the end in the second dimension
x2[0,-1]

47

You can also create 3 dimensional arrays (or any number of dimensions, really) by extending same syntax:

In [17]:
x3 = np.random.randint(10,size=(3,5,7))
x3

array([[[1, 2, 9, 1, 1, 8, 2],
        [1, 3, 3, 7, 7, 8, 3],
        [9, 5, 5, 7, 8, 7, 6],
        [6, 7, 0, 0, 2, 1, 3],
        [7, 1, 7, 7, 7, 0, 0]],

       [[9, 9, 5, 4, 8, 7, 8],
        [5, 3, 8, 1, 2, 7, 2],
        [8, 6, 3, 9, 3, 3, 0],
        [7, 5, 0, 5, 4, 8, 6],
        [4, 4, 3, 6, 1, 0, 6]],

       [[4, 4, 7, 1, 5, 3, 8],
        [9, 3, 5, 8, 2, 8, 0],
        [0, 8, 6, 0, 5, 8, 7],
        [7, 3, 9, 3, 1, 0, 8],
        [4, 4, 0, 6, 1, 7, 3]]])

In [18]:
# get a single element from 3 dimensional array
x3[1,3,4]

4

# Change value of elements in the array

You can assign individual in the array by using same syntax as getting the element followed by assignment operation (`=`). Let's remember our 2 dimensional array:

In [19]:
x2

array([[42, 32, 48, 37, 47],
       [28, 36, 43, 41, 39],
       [21, 31, 29, 13, 37]])

In [20]:
# modifying items
x2[0,0] = -666
x2[0,-1] = 999
x2

array([[-666,   32,   48,   37,  999],
       [  28,   36,   43,   41,   39],
       [  21,   31,   29,   13,   37]])

Keep in mind that NumPy arrays have fixed type, and they will not "upcast" automatically! E.g. if you try to store `float` in the array of type `int`, it will try to convert `float` to `int`:

In [21]:
x2[0,0] = 3.1415
x2

array([[  3,  32,  48,  37, 999],
       [ 28,  36,  43,  41,  39],
       [ 21,  31,  29,  13,  37]])

## Array slicing
Using *:* within brackes we can access slices of the array with the following pattern (same as `list`):
        
    x[start:stop:step]
    
If any of these are unspecified, they are assumed as following: `start=0, stop=`*size of dimension*`, step=1`

> `np.random.rand` creates uniform random numbers from 0 to 1

In [22]:
x = np.random.rand(50)
x

array([0.85327747, 0.93448524, 0.74072625, 0.28279838, 0.24394755,
       0.4228865 , 0.22602792, 0.76796705, 0.76799699, 0.86461608,
       0.45402648, 0.00244675, 0.93406553, 0.15594633, 0.01097102,
       0.49716417, 0.73698394, 0.69290145, 0.22093281, 0.6160412 ,
       0.78751804, 0.2923685 , 0.69217394, 0.16660212, 0.96843923,
       0.03480776, 0.24271733, 0.54464918, 0.17936545, 0.64589122,
       0.01291803, 0.58934853, 0.2233101 , 0.93745512, 0.34702633,
       0.0856212 , 0.95238259, 0.99235952, 0.26049627, 0.24667288,
       0.80172547, 0.01950657, 0.13298714, 0.68347658, 0.1617023 ,
       0.79454834, 0.19262115, 0.85027146, 0.6164468 , 0.74654733])

**Note**: don't confuse: this is not a 2 dimensional array, it is 1 dimensional array with 50 number in it, and it is just displayed as a 2d table for convenience. The following is how a 2 dimensional array of the same size would look like, note that each line has its own brakets (it is basically array of arrays):

In [23]:
np.random.rand(10,5)

array([[0.77140785, 0.68451117, 0.40185905, 0.73828191, 0.68036497],
       [0.76270652, 0.32974969, 0.69069238, 0.66982167, 0.4360851 ],
       [0.19953394, 0.25384876, 0.00899703, 0.32246468, 0.20356241],
       [0.56221178, 0.95976037, 0.77443959, 0.51972855, 0.67903335],
       [0.66286801, 0.56727447, 0.34316646, 0.12510377, 0.25135067],
       [0.20865799, 0.26309506, 0.11612746, 0.6921969 , 0.94684806],
       [0.33724928, 0.60988033, 0.84725721, 0.78635921, 0.89180201],
       [0.7727183 , 0.62998669, 0.58094715, 0.87058508, 0.32788016],
       [0.15248749, 0.89075764, 0.91113651, 0.06443839, 0.67522021],
       [0.27859241, 0.29902909, 0.29089561, 0.54698332, 0.25045062]])

Let get back to our 1d array with 50 numbers and look at slicing:

In [24]:
x = np.random.rand(50)
x

array([0.87161052, 0.14803519, 0.63575087, 0.087833  , 0.95611838,
       0.95486976, 0.05985936, 0.34059389, 0.42984314, 0.84703061,
       0.98996125, 0.34650466, 0.54684004, 0.90308737, 0.00833628,
       0.52795916, 0.0849075 , 0.01865644, 0.99401179, 0.68730084,
       0.3154464 , 0.9040306 , 0.17285032, 0.61847886, 0.50272755,
       0.83230226, 0.78861075, 0.36633974, 0.93193336, 0.9267869 ,
       0.74962359, 0.33381776, 0.5165517 , 0.83598746, 0.3763199 ,
       0.18674088, 0.36272191, 0.06861363, 0.35021665, 0.90778551,
       0.95834415, 0.13923709, 0.66817774, 0.067392  , 0.657958  ,
       0.13462399, 0.45315583, 0.36281752, 0.14789674, 0.70206082])

In [25]:
# first 10 elements
x[:10]

array([0.87161052, 0.14803519, 0.63575087, 0.087833  , 0.95611838,
       0.95486976, 0.05985936, 0.34059389, 0.42984314, 0.84703061])

In [26]:
# from 5th to 10th element
x[4:10]

array([0.95611838, 0.95486976, 0.05985936, 0.34059389, 0.42984314,
       0.84703061])

In [27]:
# from 11th element until the end
x[10:]

array([0.98996125, 0.34650466, 0.54684004, 0.90308737, 0.00833628,
       0.52795916, 0.0849075 , 0.01865644, 0.99401179, 0.68730084,
       0.3154464 , 0.9040306 , 0.17285032, 0.61847886, 0.50272755,
       0.83230226, 0.78861075, 0.36633974, 0.93193336, 0.9267869 ,
       0.74962359, 0.33381776, 0.5165517 , 0.83598746, 0.3763199 ,
       0.18674088, 0.36272191, 0.06861363, 0.35021665, 0.90778551,
       0.95834415, 0.13923709, 0.66817774, 0.067392  , 0.657958  ,
       0.13462399, 0.45315583, 0.36281752, 0.14789674, 0.70206082])

>**Note**: If you were doubting convenience of zero-indexing (like I was in the beginnig), perhaps this is a time to re-evaluate. Note that with zero indexing and not including end-point in the slice, it is very easy to tell how many elements you will get in the output when you slice: `x[:5]` will give you exactly `5` elements. `x[4:10]` will give exactly `10-4 = 6` elements. There is no uncertainty there, it is all very clear and consistent. For comparison, in MATLAB `x(1:5)` gives you `5-1+1 = 5` elements and `x(2:5)` gives you `5-2+1 = 4` elements.

>Another strength of this approach is that if you want to extract consequtive intervals, you end one and start next one with the same index, i.e. `x[:5]` and `x[5:]` are *non-overlaping* consequtive intervals. For comparison, in MATLAB this would become `x(1:5)` and `x(6:end)`.

>And this is only beginning, in fact any calculations on intervals are immensely simplified in zero-indexing and not including the end-point in the slice.

Some more slicing examples:

In [28]:
# every second element
x[::2]

array([0.87161052, 0.63575087, 0.95611838, 0.05985936, 0.42984314,
       0.98996125, 0.54684004, 0.00833628, 0.0849075 , 0.99401179,
       0.3154464 , 0.17285032, 0.50272755, 0.78861075, 0.93193336,
       0.74962359, 0.5165517 , 0.3763199 , 0.36272191, 0.35021665,
       0.95834415, 0.66817774, 0.657958  , 0.45315583, 0.14789674])

In [29]:
# every third element
x[::3]

array([0.87161052, 0.087833  , 0.05985936, 0.84703061, 0.54684004,
       0.52795916, 0.99401179, 0.9040306 , 0.50272755, 0.36633974,
       0.74962359, 0.83598746, 0.36272191, 0.90778551, 0.66817774,
       0.13462399, 0.14789674])

In [30]:
# reverse array
x = np.arange(10)
print(x)
print(x[::-1])

[0 1 2 3 4 5 6 7 8 9]
[9 8 7 6 5 4 3 2 1 0]


You can combine slicing in one dimension with precise indexing in another:

In [31]:
x2

array([[  3,  32,  48,  37, 999],
       [ 28,  36,  43,  41,  39],
       [ 21,  31,  29,  13,  37]])

In [32]:
# access third column
x2[:,2]

array([48, 43, 29])

> **Pro-tip**: When slicing an array, it is useful to keep in mind that by default the resulting sub-array is not a separate entity in memory, but is actually accessing the memory location of the original array. In programming terms we can say that it is *view* on the same object, not a copy. Here is a simple example:

In [33]:
x2 = np.random.randint(0, 10, size=(3,5))
x2

array([[1, 9, 7, 8, 6],
       [3, 0, 2, 5, 9],
       [9, 7, 4, 8, 5]])

In [34]:
# get first two elements from both dimensions
x2_sub = x2[:2,:2]
x2_sub

array([[1, 9],
       [3, 0]])

In [35]:
# modify an element in the new array
x2_sub[1,1] = 999
x2_sub

array([[  1,   9],
       [  3, 999]])

In [36]:
# see that the original array also got modified
x2

array([[  1,   9,   7,   8,   6],
       [  3, 999,   2,   5,   9],
       [  9,   7,   4,   8,   5]])

>This behavior is very useful for working with data, because it saves you a lot of memory and speed for useless copying. If you need to make a copy of the array, you must do it explicitly:

In [37]:
# make a copy
x2_sub = x2[:2,:2].copy()

In [38]:
# modify a copy and verify that the original array is intact
x2_sub[1,1] = -666
print(x2_sub)
print()
print(x2)

[[   1    9]
 [   3 -666]]

[[  1   9   7   8   6]
 [  3 999   2   5   9]
 [  9   7   4   8   5]]


# Operations on arrays

We use arrays to speed up computations. The key thing to understand here is that when you make an operation on each element of an array or a list separately, each object has to be *dynamically typed*: during the execusion for each element Python core has to look up at the type of the element (whether it is `int`, `float`, `str`, etc) to see what "flavour" of the function to apply to the element. For example, in terms of precise operations on memory, convering `float` to `int` -- `int(5.6)` --  is not the same as converting `str` to `int` -- `int('5')` -- therefore before applying a particular routine to the object, Python must check the type. This is slow. Consider the following piece of code, as an example:

In [39]:
def compute_reciprocals(values):
    
    # preallocate array for reciprocals, same length as `values` input
    output = np.empty(len(values))

    # compute reciprocal of each element
    for i in range(len(values)):
        output[i] = 1.0 / values[i] # each time this division runs, Python must check types behind the scenes

    return output

# example use
values = np.arange(1, 10)
print(values)
compute_reciprocals(values)

[1 2 3 4 5 6 7 8 9]


array([1.        , 0.5       , 0.33333333, 0.25      , 0.2       ,
       0.16666667, 0.14285714, 0.125     , 0.11111111])

This works fine. Now let's see how much time it takes to run this function on an array of 1 million integers. We use `%timeit`, which estimates the time it takes to run the function you wrote afterwards:

In [40]:
big_array = np.random.randint(1, 100, size=1000000)
%timeit compute_reciprocals(big_array)

1.64 s ± 59.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Now we do the same thing, but instead we just divide `1` over our `big_array`. `numpy` automatically assumes that we want to divide `1` by each element of the array. Moreover, it has more efficient ways of doing it for us:

In [41]:
%timeit (1/big_array)

1.38 ms ± 84.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


So we did the same thing two times in two different ways and checked the time it takes. Results will wary based on the current state of your computer and its CPU speed, but here is what I got: for `compute_reciprocals` function I got `2.87 s`, and for `(1/big_array)` I got `6.17 ms`. This is almost 500 fold difference in run time! Imagine that you had a script using `numpy` which ran for 10 seconds. If you had written it in a wrong way using loops, it would take almost 1.5 hours to run!

>**Pro-tip**: although we used `/` to divide `1/big_array`, this operator `/` is actually a shortcut for a function `np.divide':

In [42]:
np.divide(1,big_array)

array([0.01136364, 0.02380952, 0.25      , ..., 0.01075269, 0.2       ,
       0.05882353])

>All `numpy` operators have functions associated with them, here they are:

| Operator	    | Equivalent ufunc    | Description                           |
|---------------|---------------------|---------------------------------------|
|``+``          |``np.add``           |Addition (e.g., ``1 + 1 = 2``)         |
|``-``          |``np.subtract``      |Subtraction (e.g., ``3 - 2 = 1``)      |
|``-``          |``np.negative``      |Unary negation (e.g., ``-2``)          |
|``*``          |``np.multiply``      |Multiplication (e.g., ``2 * 3 = 6``)   |
|``/``          |``np.divide``        |Division (e.g., ``3 / 2 = 1.5``)       |
|``//``         |``np.floor_divide``  |Floor division (e.g., ``3 // 2 = 1``)  |
|``**``         |``np.power``         |Exponentiation (e.g., ``2 ** 3 = 8``)  |
|``%``          |``np.mod``           |Modulus/remainder (e.g., ``9 % 4 = 1``)|

> Why would you want to use the full function notation instead of using an operator? There are some curcumstances when using full function notation will give you more flexibility. If you do a lot of crunching on very large numerical datasets and experience problems with speed and/or memory, certainly take a look at the [NumPy](http://www.numpy.org) documentation (especially at the [ufunc](https://docs.scipy.org/doc/numpy-1.10.0/reference/ufuncs.html) section) and [SciPy](http://www.scipy.org) documentation.

# Aggregator functions

Functions which reduce an array (or a dimension of an array) to a single value are called aggregator functions. Some of the most useful include `sum`, `min`, `max`, `mean`, `median`, `std`, etc. Python has in-built versions of some of these functions, but `numpy` versions are much faster and you should be always using them:

In [43]:
big_array = np.random.rand(1000000)
%timeit sum(big_array)
%timeit np.sum(big_array)

77.7 ms ± 5.33 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
384 µs ± 10.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Most of the aggregator function include a sister-function with `nan` prefix, which does the same, but ignores `NaN` (stands for *Not a Number*) elements. `NaN` is usually used as a placeholder for missing data, so these functions are very useful for working with data. We will revisit this in the future lesson.

The following table provides a list of useful aggregation functions available in NumPy:

|Function name      |   NaN-ignoring version  | Description                                   |
|-------------------|---------------------|-----------------------------------------------|
| ``np.sum``        | ``np.nansum``       | Compute sum of elements                       |
| ``np.prod``       | ``np.nanprod``      | Compute product of elements                   |
| ``np.mean``       | ``np.nanmean``      | Compute median of elements                    |
| ``np.std``        | ``np.nanstd``       | Compute standard deviation                    |
| ``np.var``        | ``np.nanvar``       | Compute variance                              |
| ``np.min``        | ``np.nanmin``       | Find minimum value                            |
| ``np.max``        | ``np.nanmax``       | Find maximum value                            |
| ``np.argmin``     | ``np.nanargmin``    | Find index of minimum value                   |
| ``np.argmax``     | ``np.nanargmax``    | Find index of maximum value                   |
| ``np.median``     | ``np.nanmedian``    | Compute median of elements                    |
| ``np.percentile`` | ``np.nanpercentile``| Compute rank-based statistics of elements     |
| ``np.any``        | N/A                 | Evaluate whether any elements are True        |
| ``np.all``        | N/A                 | Evaluate whether all elements are True        |

We won't discuss each in detail, but feel free to try them for youself.

Just to mention 2 things about aggregation functions. 

**First**, some of them (`sum`, `min`, `max` and some others) can be accessed via method notation, like so:

In [44]:
# print min, max and sum of the array
print('Min:', big_array.min())
print('Max:', big_array.max())
print('Sum:', big_array.sum())

Min: 3.5411831977860686e-06
Max: 0.9999998888474394
Sum: 499860.1415110606


You will notice this pattern in Python further, it is quite frequent that you can do something using a function, or a method. Usually there is no difference, but sometimes one approach offers some advantage in terms of code length and/or clarity.

And **second**, for multidimensional arrays, you can specify `axis` parameter to make aggregation only over a specific axis. By default, they will aggregate over all the array:

In [45]:
multi_dim_array = np.random.randint(100, size=(5,10))
multi_dim_array

array([[ 6, 19, 63, 71, 95, 56, 61, 77,  7, 90],
       [18, 49, 42, 98, 64, 62, 69, 45, 32, 82],
       [69, 71, 79, 63, 69, 10, 82,  7, 22, 81],
       [43, 76, 40, 87, 14,  0, 76, 32, 93, 74],
       [95, 64, 72, 47, 34,  5, 53, 36, 45, 49]])

In [46]:
# default behavior gives maximum of all elements of the array 
np.max(multi_dim_array)

98

In [47]:
# specifying axis gives you control over which dimension is aggregated;
# in this particular case, the function will give max of every column 
np.max(multi_dim_array, axis=0)

array([95, 76, 79, 98, 95, 62, 82, 77, 93, 90])

# Troubles with floats
Example:

In [48]:
x = 0.1 + 0.2
y = x == 0.3
y

False

Why this happens? If you can spare 9 minutes of your time, I suggest you to watch a great explanation on <a href="https://www.youtube.com/watch?v=PZRI1IfStY0">Computerphile</a> youtube channel. You will understand how floating numbers are stored and why the floating arithmetic is not precise.

I will attempt to explain it here in a nutshell. The problem is more evident if we just look at what `0.1 + 0.2` gives us:

In [49]:
0.1+0.2

0.30000000000000004

As you can see, there is a marginal error in the end. Why is it there? The answer has to do with how numbers (in particular, `float` type) are stored in memory: they can only store a certain number of *significant digits*. In situations when the precision requires an infinite number of repeating digits, it will fail. It is easier to understand with an analogy between fractions and decimals.

Think about a fraction `1/3`. If you try to write it as a decimal, you'd have to write `0.33333333...` and so on. But what if you were able to only store 5 significant digits after `0.`? You'd have to write `0.33333`, that's the best approximation you can do. Now if you try to do arithmetic with `1/3`, let's say `1/3 + 1/3 + 1/3 = 3`. However, if you try to do it in decimal where you can only store 5 significant digits, you'd get `0.99999`.

A similar thing happens in the computers when you write any decimals (like `0.1` and `0.2`), because computers cannot store decimal notation in memory, they have to translate them to *binary*. And sometimes with this translation you get a repeating pattern, just like with `0.33333...`. But computers cannot store infinite numbers, they have to cut it after some digits. This introduces errors when working with `float`, and it will happen in any language, because it is a fundamental property of computers.

Based on these difficulties, there are 2 fundamental principles one has to keep in mind when working with `float` type.

**First**: Never compare equality of floats, this is exactly same mistake which we did in the script from the survey:
    
    x = 0.1 + 0.2
    y = x == 0.3
    
The result will sometimes be `True`, sometimes `False`, based on which numbers exactly you try, but the point is that it is *inconsistent*, you cannot trust it. It is usually fine to compare which float is larger though. (**Pro-tip**: if you ever do need to compare equality of floats and cannot get around it, there is function called <a href="https://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.isclose.html">`isclose`</a> in the `numpy` package, which will compare two things up to a specified tolerance).

**Second**: Be very careful when doing summation or difference between floats of vastly different order: the smaller ones can get lost in the noise. Consider the following example, where we add `0.001` one thousand times to a large number $10^{13}$, and expect to get $10^{13}+1$:

In [ ]:
a = 10.0**13
for i in range(1000):
    a = a + 0.001
a

As you can see, the error is almost the same size as the number we added.

**Pro-tip**: if you even need to do very precise operations with `float` type, check out <a href="https://docs.python.org/2/library/decimal.html">`decimal`</a> module in Python)

If you want to dig a little deeper on that, check out this really great page: http://floating-point-gui.de/